// reshaped/dist/styles/resolvers/margin/index.js
import { responsiveVariables } from "../../../utilities/props.js";
import "./margin.css";
const margin = (value) => {
    if (!value)
        return {};
    return { variables: responsiveVariables("--rs-m", value) };
};
export const marginTop = (value) => {
    if (value === undefined)
        return {};
    return { variables: responsiveVariables("--rs-m-top", value) };
};
export const marginBottom = (value) => {
    if (value === undefined)
        return {};
    return { variables: responsiveVariables("--rs-m-bottom", value) };
};
export const marginStart = (value) => {
    if (value === undefined)
        return {};
    return { variables: responsiveVariables("--rs-m-start", value) };
};
export const marginEnd = (value) => {
    if (value === undefined)
        return {};
    return { variables: responsiveVariables("--rs-m-end", value) };
};
export const marginInline = (value) => {
    if (value === undefined)
        return {};
    return { variables: responsiveVariables("--rs-m-inline", value) };
};
export const marginBlock = (value) => {
    if (value === undefined)
        return {};
    return { variables: responsiveVariables("--rs-m-block", value) };
};
export default margin;
