// reshaped/dist/hooks/_private/useDrag.js
"use client";
import React from "react";
import { disableUserSelect, enableUserSelect } from "../../utilities/dom/index.js";
import { disableScroll, enableScroll } from "../../utilities/scroll/index.js";
import useToggle from "../useToggle.js";
import useHotkeys from "../useHotkeys.js";
import useHandlerRef from "../useHandlerRef.js";
import * as keys from "../../constants/keys.js";
const useDrag = (cb, options) => {
    const { disabled, containerRef: passedContainerRef, orientation = "all" } = options || {};
    const cbRef = useHandlerRef(cb);
    const toggle = useToggle();
    const triggerRef = React.useRef(null);
    const internalContainerRef = React.useRef(null);
    const containerRef = passedContainerRef || internalContainerRef;
    const triggerCompensationRef = React.useRef({ x: 0, y: 0 });
    const isVertical = orientation === "vertical" || orientation === "all";
    const isHorizontal = orientation === "horizontal" || orientation === "all";
    const handleKeyboard = (x, y) => {
        const triggerEl = triggerRef.current;
        if (!triggerEl)
            return;
        const container = containerRef.current ?? document.body;
        const containerRect = container.getBoundingClientRect();
        const triggerRect = triggerEl?.getBoundingClientRect();
        const nextArgs = { x: 0, y: 0, triggerX: 0, triggerY: 0 };
        if (isVertical) {
            const relativeY = Math.round(triggerRect.y) - containerRect.y + y;
            nextArgs.y = Math.max(0, Math.min(relativeY, containerRect.height - triggerRect.height));
            nextArgs.triggerY = triggerRect.y - containerRect.y;
        }
        if (isHorizontal) {
            const relativeX = Math.round(triggerRect.x) - containerRect.x + x;
            nextArgs.x = Math.max(0, Math.min(relativeX, containerRect.width - triggerRect.width));
            nextArgs.triggerX = triggerRect.x - containerRect.x;
        }
        cb(nextArgs);
    };
    useHotkeys({
        [keys.LEFT]: () => isHorizontal && handleKeyboard(-20, 0),
        [keys.RIGHT]: () => isHorizontal && handleKeyboard(20, 0),
        [keys.UP]: () => isVertical && handleKeyboard(0, -20),
        [keys.DOWN]: () => isVertical && handleKeyboard(0, 20),
    }, [], {
        ref: triggerRef,
        preventDefault: true,
        disabled,
    });
    React.useEffect(() => {
        const triggerEl = triggerRef.current;
        if (!triggerEl)
            return;
        if (!toggle.active)
            return;
        const handleDrag = (event) => {
            const resolvedEvent = event instanceof MouseEvent ? event : event.changedTouches[0];
            const container = containerRef.current ?? document.body;
            const containerRect = container.getBoundingClientRect();
            const triggerRect = triggerEl.getBoundingClientRect();
            const triggerX = resolvedEvent.clientX - containerRect.x;
            const triggerY = resolvedEvent.clientY - containerRect.y;
            // Calculate position relative to the container
            const relativeX = triggerX - triggerCompensationRef.current.x;
            const relativeY = triggerY - triggerCompensationRef.current.y;
            cbRef.current?.({
                x: isHorizontal
                    ? Math.max(0, Math.min(relativeX, containerRect.width - triggerRect.width))
                    : 0,
                y: isVertical
                    ? Math.max(0, Math.min(relativeY, containerRect.height - triggerRect.height))
                    : 0,
                triggerX: triggerRect.x - containerRect.x,
                triggerY: triggerRect.y - containerRect.y,
            });
        };
        const handleDragEnd = () => {
            triggerCompensationRef.current = { x: 0, y: 0 };
            toggle.deactivate();
            enableUserSelect();
            enableScroll();
        };
        document.addEventListener("touchmove", handleDrag, { passive: true });
        document.addEventListener("touchend", handleDragEnd, { passive: true });
        document.addEventListener("mousemove", handleDrag, { passive: true });
        document.addEventListener("mouseup", handleDragEnd, { passive: true });
        return () => {
            document.removeEventListener("touchmove", handleDrag);
            document.removeEventListener("touchend", handleDragEnd);
            document.removeEventListener("mousemove", handleDrag);
            document.removeEventListener("mouseup", handleDragEnd);
        };
    }, [toggle, isHorizontal, isVertical, containerRef, cbRef]);
    React.useEffect(() => {
        const triggerEl = triggerRef.current;
        if (!triggerEl || disabled)
            return;
        const handleStart = (event) => {
            const resolvedEvent = event instanceof MouseEvent ? event : event.changedTouches[0];
            // Find the coordinate of the event inside the trigger
            const triggerRect = triggerEl.getBoundingClientRect();
            triggerCompensationRef.current = {
                x: resolvedEvent.clientX - triggerRect.x,
                y: resolvedEvent.clientY - triggerRect.y,
            };
            toggle.activate();
            disableUserSelect();
            disableScroll();
        };
        triggerEl.addEventListener("touchstart", handleStart, { passive: true });
        triggerEl.addEventListener("mousedown", handleStart, { passive: true });
        return () => {
            triggerEl.removeEventListener("touchstart", handleStart);
            triggerEl.removeEventListener("mousedown", handleStart);
        };
    }, [toggle, disabled]);
    return { ref: triggerRef, containerRef, active: toggle.active };
};
export default useDrag;
