// reshaped/dist/components/Theme/tests/Theme.stories.js
import { expect, userEvent } from "storybook/test";
import Card from "../../Card/index.js";
import Button from "../../Button/index.js";
import View from "../../View/index.js";
import MenuItem from "../../MenuItem/index.js";
import Theme, { useTheme } from "../index.js";
import { Example } from "../../../utilities/storybook/index.js";
import Popover from "../../Popover/index.js";
export default {
    title: "Utility components/Theme",
    component: Theme,
    parameters: {
        iframe: {
            url: "https://reshaped.so/docs/utilities/theme",
        },
        chromatic: { disableSnapshot: true },
    },
};
export const scoped = {
    name: "scoped",
    render: () => (<Example>
			<Example.Item title="scoped, single">
				<Theme name="reshaped">
					<Card attributes={{ "data-testid": "test-id" }}>
						<Button color="primary">Action</Button>
					</Card>
				</Theme>
			</Example.Item>

			<Example.Item title="scoped, multiple">
				<Theme name={["reshaped", "figma"]}>
					<Card attributes={{ "data-testid": "test-id-multi" }}>
						<View direction="row" gap={4}>
							<Button color="primary">Action</Button>

							<Popover>
								<Popover.Trigger>
									{(attributes) => <Button attributes={attributes}>Popover</Button>}
								</Popover.Trigger>
								<Popover.Content>Content</Popover.Content>
							</Popover>
						</View>
					</Card>
				</Theme>
			</Example.Item>
		</Example>),
    play: ({ canvas }) => {
        const root = canvas.getByTestId("test-id").parentNode;
        const rootMultiple = canvas.getByTestId("test-id-multi").parentNode;
        expect(root).toHaveAttribute("data-rs-theme", "reshaped");
        expect(root).toHaveAttribute("data-rs-color-mode", "dark");
        expect(rootMultiple).toHaveAttribute("data-rs-theme", " reshaped figma ");
        expect(root).toHaveAttribute("data-rs-color-mode", "dark");
    },
};
export const light = {
    name: "light",
    render: () => (<Theme name="reshaped" colorMode="light">
			<Card attributes={{ "data-testid": "test-id" }}>Content</Card>
		</Theme>),
    play: ({ canvas }) => {
        const root = canvas.getByTestId("test-id").parentNode;
        expect(root).toHaveAttribute("data-rs-theme", "reshaped");
        expect(root).toHaveAttribute("data-rs-color-mode", "light");
    },
};
export const dark = {
    name: "dark",
    render: () => (<Theme name="reshaped" colorMode="dark">
			<Card attributes={{ "data-testid": "test-id" }}>Content</Card>
		</Theme>),
    play: ({ canvas }) => {
        const root = canvas.getByTestId("test-id").parentNode;
        expect(root).toHaveAttribute("data-rs-theme", "reshaped");
        expect(root).toHaveAttribute("data-rs-color-mode", "dark");
    },
};
export const inherited = {
    name: "inherited",
    render: () => (<Theme name="reshaped">
			<Card attributes={{ "data-testid": "test-id" }}>Content</Card>
		</Theme>),
    play: ({ canvas }) => {
        const root = canvas.getByTestId("test-id").parentNode;
        expect(root).toHaveAttribute("data-rs-theme", "reshaped");
        // Our storybook setup uses dark by default
        expect(root).toHaveAttribute("data-rs-color-mode", "dark");
    },
};
export const inverted = {
    name: "inverted",
    render: () => (<Theme name="reshaped" colorMode="inverted">
			<Card attributes={{ "data-testid": "test-id" }}>Content</Card>
		</Theme>),
    play: ({ canvas }) => {
        const root = canvas.getByTestId("test-id").parentNode;
        expect(root).toHaveAttribute("data-rs-theme", "reshaped");
        expect(root).toHaveAttribute("data-rs-color-mode", "light");
    },
};
export const controlled = {
    name: "controlled",
    render: () => {
        const Internal = () => {
            const { setTheme, theme } = useTheme();
            return (<Button color="primary" onClick={() => setTheme("slate")}>
					Switch to slate
				</Button>);
        };
        return (<div data-testid="root">
				<Theme name="reshaped">
					<Internal />
				</Theme>
			</div>);
    },
    play: async ({ canvas }) => {
        const root = canvas.getByTestId("root").firstChild;
        const trigger = canvas.getAllByRole("button")[0];
        expect(root).toHaveAttribute("data-rs-theme", "reshaped");
        expect(root).toHaveAttribute("data-rs-color-mode", "dark");
        await userEvent.click(trigger);
        expect(root).toHaveAttribute("data-rs-theme", "reshaped");
        expect(root).toHaveAttribute("data-rs-color-mode", "dark");
    },
};
export const uncontrolled = {
    name: "uncontrolled",
    render: () => {
        const Internal = () => {
            const { setTheme } = useTheme();
            return (<Button color="primary" onClick={() => setTheme("slate")}>
					Switch to slate
				</Button>);
        };
        return (<div data-testid="root">
				<Theme defaultName="reshaped">
					<Internal />
				</Theme>
			</div>);
    },
    play: async ({ canvas }) => {
        const root = canvas.getByTestId("root").firstChild;
        const trigger = canvas.getAllByRole("button")[0];
        expect(root).toHaveAttribute("data-rs-theme", "reshaped");
        expect(root).toHaveAttribute("data-rs-color-mode", "dark");
        await userEvent.click(trigger);
        expect(root).toHaveAttribute("data-rs-theme", "slate");
        expect(root).toHaveAttribute("data-rs-color-mode", "dark");
    },
};
export const disabledTransition = {
    name: "disabled transitions",
    render: () => {
        const { invertColorMode } = useTheme();
        return (<Example>
				<Example.Item title="should have no transitions while switching color mode">
					<View gap={3}>
						<Button onClick={invertColorMode}>Invert mode</Button>

						<MenuItem selected>Test transition</MenuItem>

						<Card>Default card</Card>

						<Theme colorMode="inverted">
							<Card>Inverted card</Card>
						</Theme>
					</View>
				</Example.Item>
			</Example>);
    },
};
