// reshaped/dist/themes/_generator/tokens/css.js
import { camelToKebab } from "../../../utilities/string.js";
export const getVariableName = (tokenName, tokenType) => {
    const value = ["rs", tokenType && camelToKebab(tokenType), camelToKebab(tokenName)]
        .filter(Boolean)
        .join("-");
    return `--${value}`;
};
export const createVariable = (token) => {
    const variableName = getVariableName(token.name, token.tokenType);
    return `${variableName}: ${token.value};`;
};
export const createMedia = (token) => `@custom-media ${getVariableName(token.name, token.tokenType)} ${token.value};`;
export const variablesTemplate = (themeName, tokens) => {
    const filteredTokens = {
        root: [],
        light: [],
        dark: [],
    };
    tokens.forEach((token) => {
        if (token.mode === "light")
            filteredTokens.light.push(token);
        if (token.mode === "dark")
            filteredTokens.dark.push(token);
        if (!token.mode)
            filteredTokens.root.push(token);
    });
    let code = "";
    Object.entries(filteredTokens).forEach(([type, tokens]) => {
        const selector = type === "root"
            ? `[data-rs-theme~="${themeName}"]`
            : `[data-rs-theme~="${themeName}"][data-rs-color-mode="${type}"]`;
        if (!tokens.length)
            return;
        code += `
			${selector} {
					${tokens.map((token) => createVariable(token)).join("\n")}
			}
		`;
    });
    return code;
};
export const mediaTemplate = (tokens) => `
	${tokens.map((token) => createMedia(token)).join("\n")}
`;
