// reshaped/dist/components/Alert/tests/Alert.stories.js
import { Example, Placeholder } from "../../../utilities/storybook/index.js";
import Alert from "../index.js";
import Link from "../../Link/index.js";
import IconZap from "../../../icons/Zap.js";
import { expect } from "storybook/test";
export default {
    title: "Components/Alert",
    component: Alert,
    parameters: {
        iframe: {
            url: "https://reshaped.so/docs/components/alert",
        },
    },
};
export const color = {
    name: "color",
    render: () => (<Example>
			{["neutral", "primary", "critical", "positive", "warning"].map((color) => (<Example.Item title={`color: ${color}`} key={color}>
					<Alert color={color} title="Alert title goes here" icon={IconZap} actionsSlot={<>
								<Link variant="plain" color={color === "neutral" ? "primary" : color} onClick={() => { }}>
									View now
								</Link>
								<Link variant="plain" color={color === "neutral" ? "primary" : color} onClick={() => { }}>
									Dismiss
								</Link>
							</>}>
						Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum
						has been the industry's standard
					</Alert>
				</Example.Item>))}
		</Example>),
};
export const inline = {
    name: "inline",
    render: () => (<Example>
			<Example.Item title="inline: true">
				<Alert inline title="Alert title goes here" icon={IconZap} actionsSlot={<>
							<Link variant="plain" onClick={() => { }}>
								View now
							</Link>
							<Link variant="plain" onClick={() => { }}>
								Dismiss
							</Link>
						</>}>
					Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has
					been the industry's standard
				</Alert>
			</Example.Item>
		</Example>),
};
export const bleed = {
    name: "bleed",
    render: () => (<Example>
			<Example.Item title="bleed: 4">
				<Alert bleed={4} icon={IconZap}>
					Content
				</Alert>
			</Example.Item>
			<Example.Item title="bleed: [s] 4, [m+] 0">
				<Alert bleed={{ s: 4, m: 0 }} icon={IconZap}>
					Content
				</Alert>
			</Example.Item>
		</Example>),
};
export const className = {
    name: "className, attributes",
    render: () => (<div data-testid="root">
			<Alert className="test-classname" attributes={{ id: "test-id" }}>
				<Placeholder />
			</Alert>
		</div>),
    play: async ({ canvas }) => {
        const root = canvas.getByTestId("root").firstChild;
        expect(root).toHaveClass("test-classname");
        expect(root).toHaveAttribute("id", "test-id");
    },
};
