// reshaped/dist/hooks/tests/useHandlerRef.stories.js
import React from "react";
import { expect, fn, userEvent } from "storybook/test";
import Button from "../../components/Button/index.js";
import View from "../../components/View/index.js";
import useHandlerRef from "../useHandlerRef.js";
export default {
    title: "Hooks/useHandlerRef",
    parameters: {
        chromatic: { disableSnapshot: true },
    },
};
const Component = (props) => {
    const { onEffect, count } = props;
    const onEffectRef = useHandlerRef(onEffect);
    React.useEffect(() => {
        onEffectRef.current();
    }, [onEffectRef]);
    return <>Counter: {count}</>;
};
export const base = {
    name: "base",
    args: {
        handleEffect: fn(),
    },
    render: (args) => {
        const [count, setCount] = React.useState(0);
        // Creating a new handler on each render
        const handleEffect = () => {
            args.handleEffect(0);
        };
        return (<View gap={4}>
				<Button onClick={() => setCount((prev) => prev + 1)}>Increase count</Button>
				<Component onEffect={handleEffect} count={count}/>
			</View>);
    },
    play: async ({ canvas, args }) => {
        const button = canvas.getAllByRole("button")[0];
        // mount, triggerred twice in dev mode
        expect(args.handleEffect).toHaveBeenCalledTimes(2);
        await userEvent.click(button);
        // Rerendering the component doesn't re-trigger the effect
        expect(args.handleEffect).toHaveBeenCalledTimes(2);
    },
};
