// reshaped/dist/hooks/tests/useOnClickOutside.stories.js
import React from "react";
import { expect, fn, userEvent } from "storybook/test";
import Button from "../../components/Button/index.js";
import useOnClickOutside from "../useOnClickOutside.js";
import View from "../../components/View/index.js";
export default {
    title: "Hooks/useOnClickOutside",
    parameters: {
        chromatic: { disableSnapshot: true },
    },
};
export const base = {
    name: "base",
    args: {
        handleOutsideClick: fn(),
    },
    render: (args) => {
        const ref = React.useRef(null);
        const [target, setTarget] = React.useState(null);
        useOnClickOutside([ref], () => {
            args.handleOutsideClick();
            setTarget("outside");
        });
        return (<View gap={4} align="start">
				<Button attributes={{ ref }} onClick={() => setTarget("inside")}>
					Trigger
				</Button>
				{target && `Clicked ${target}`}
			</View>);
    },
    play: async ({ canvas, args }) => {
        const button = canvas.getAllByRole("button")[0];
        await userEvent.click(button);
        expect(args.handleOutsideClick).not.toHaveBeenCalled();
        await userEvent.click(document.body);
        expect(args.handleOutsideClick).toHaveBeenCalledTimes(1);
        expect(args.handleOutsideClick).toHaveBeenCalledWith();
    },
};
export const refs = {
    name: "multiple refs",
    args: {
        handleOutsideClick: fn(),
    },
    render: (args) => {
        const ref = React.useRef(null);
        const ref2 = React.useRef(null);
        useOnClickOutside([ref, ref2], () => {
            args.handleOutsideClick();
        });
        return (<View gap={4} align="start">
				<Button attributes={{ ref }}>Trigger</Button>
				<Button attributes={{ ref: ref2 }}>Trigger 2</Button>
			</View>);
    },
    play: async ({ canvas, args }) => {
        const [button, button2] = canvas.getAllByRole("button");
        await userEvent.click(button);
        expect(args.handleOutsideClick).not.toHaveBeenCalled();
        await userEvent.click(button2);
        expect(args.handleOutsideClick).not.toHaveBeenCalled();
        await userEvent.click(document.body);
        expect(args.handleOutsideClick).toHaveBeenCalledTimes(1);
        expect(args.handleOutsideClick).toHaveBeenCalledWith();
    },
};
export const disabled = {
    name: "disabled",
    args: {
        handleOutsideClick: fn(),
    },
    render: (args) => {
        const ref = React.useRef(null);
        useOnClickOutside([ref], () => {
            args.handleOutsideClick();
        }, {
            disabled: true,
        });
        return (<View gap={4} align="start">
				<Button attributes={{ ref }}>Trigger</Button>
			</View>);
    },
    play: async ({ args }) => {
        await userEvent.click(document.body);
        expect(args.handleOutsideClick).not.toHaveBeenCalled();
    },
};
export const deps = {
    name: "test: handler uses latest state",
    args: {
        handleOutsideClick: fn(),
    },
    render: (args) => {
        const ref = React.useRef(null);
        const [count, setCount] = React.useState(0);
        useOnClickOutside([ref], () => {
            args.handleOutsideClick({ count });
        });
        return (<Button attributes={{ ref }} onClick={() => setCount((prev) => prev + 1)}>
				Trigger
			</Button>);
    },
    play: async ({ canvas, args }) => {
        const button = canvas.getAllByRole("button")[0];
        await userEvent.click(document.body);
        expect(args.handleOutsideClick).toHaveBeenLastCalledWith({ count: 0 });
        await userEvent.click(button);
        await userEvent.click(document.body);
        expect(args.handleOutsideClick).toHaveBeenLastCalledWith({ count: 1 });
    },
};
