// reshaped/dist/hooks/tests/useToggle.stories.js
import React from "react";
import { expect, userEvent } from "storybook/test";
import Button from "../../components/Button/index.js";
import useToggle from "../useToggle.js";
export default {
    title: "Hooks/useToggle",
    parameters: {
        chromatic: { disableSnapshot: true },
    },
};
export const toggle = {
    name: "toggle",
    render: () => {
        const { toggle, active } = useToggle();
        return (<Button onClick={() => toggle()} attributes={{ "data-active": active }}>
				{active ? "Deactivate" : "Activate"}
			</Button>);
    },
    play: async ({ canvas }) => {
        const button = canvas.getAllByRole("button")[0];
        expect(button.getAttribute("data-active")).toBe("false");
        await userEvent.click(button);
        expect(button.getAttribute("data-active")).toBe("true");
        await userEvent.click(button);
        expect(button.getAttribute("data-active")).toBe("false");
    },
};
export const activate = {
    name: "activate",
    render: () => {
        const { activate, active } = useToggle();
        return (<React.Fragment>
				<Button onClick={activate} attributes={{ "data-active": active }}>
					Activate
				</Button>
			</React.Fragment>);
    },
    play: async ({ canvas }) => {
        const button = canvas.getAllByRole("button")[0];
        expect(button.getAttribute("data-active")).toBe("false");
        await userEvent.click(button);
        expect(button.getAttribute("data-active")).toBe("true");
    },
};
export const deactivate = {
    name: "deactivate",
    render: () => {
        const { deactivate, active } = useToggle(true);
        return (<Button onClick={deactivate} attributes={{ "data-active": active }}>
				Deactivate
			</Button>);
    },
    play: async ({ canvas }) => {
        const button = canvas.getAllByRole("button")[0];
        expect(button.getAttribute("data-active")).toBe("true");
        await userEvent.click(button);
        expect(button.getAttribute("data-active")).toBe("false");
    },
};
