// reshaped/dist/components/Hotkey/tests/Hotkey.stories.js
import { Example } from "../../../utilities/storybook/index.js";
import useHotkeys from "../../../hooks/useHotkeys.js";
import View from "../../View/index.js";
import TextField from "../../TextField/index.js";
import Hotkey from "../Hotkey.js";
import { expect } from "storybook/test";
export default {
    title: "Components/Hotkey",
    component: Hotkey,
    parameters: {
        iframe: {
            url: "https://reshaped.so/docs/components/hotkey",
        },
    },
};
const Demo = () => {
    const { checkHotkeyState } = useHotkeys({
        "Meta+v": () => {
            console.log("meta v");
        },
        "Mod+v": () => {
            console.log("mod v");
        },
    });
    return <Hotkey active={checkHotkeyState("k")}>⌘K</Hotkey>;
};
export const base = () => (<Example>
		<Example.Item title="Base">
			<Demo />
		</Example.Item>
		<Example.Item title="Active">
			<Hotkey active>⌘K</Hotkey>
		</Example.Item>
		<Example.Item title="Inside input slot">
			<View width="300px">
				<TextField name="hey" endSlot={<Demo />} inputAttributes={{ "aria-label": "hotkey test" }}/>
			</View>
		</Example.Item>
	</Example>);
export const className = {
    name: "className, attributes",
    render: () => (<div data-testid="root">
			<Hotkey className="test-classname" attributes={{ id: "test-id" }}>
				⌘K
			</Hotkey>
		</div>),
    play: async ({ canvas }) => {
        const root = canvas.getByTestId("root").firstChild;
        expect(root).toHaveClass("test-classname");
        expect(root).toHaveAttribute("id", "test-id");
        expect(root?.tagName).toBe("KBD");
    },
};
